File: Data Structures & Algorithms/anagram-groups/submission-6.cpp
using namespace std;
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map< string, vector<string> > groups;

        for (size_t idx = 0; idx < strs.size(); idx++){
            string key = strs.at(idx);
            sort(key.begin(), key.end());
            groups[key].push_back(strs.at(idx));
        }

        vector< vector<string> > result;
        for (const auto& [key, val]: groups){
            result.push_back(val);
        }
        return result;
    }
};

File: Data Structures & Algorithms/string-encode-and-decode/submission-6.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        return "".join(f'{len(word)}:{word}' for word in strs)

    def decode(self, s: str) -> List[str]:

        result = []

        idx = 0
        while idx < len(s):
            end = s.index(':', idx)
            length = int(s[idx:end])
            result.append(s[end+1: end+1+length])
            idx = end + 1 + length
        
        return result
